### ISSUE TITLE
Failed to deploy function with deployer

### ISSUE THREAD

**Issue by niravnshah** [ease-lab/vhive#476]
**Describe the bug**
when deployer is run with default value in functions.json, the deployer fails to deploy the fucntions correctly. Shows following error.
```
nirav@984fee0096e3:~/vhive$ source /etc/profile && pushd ./examples/deployer && go build && popd && ./examples/deployer/deployer
~/vhive/examples/deployer ~/vhive
go: downloading github.com/sirupsen/logrus v1.8.1
go: downloading golang.org/x/sys v0.0.0-20191026070338-33540a1f6037
~/vhive
WARN[0013] Failed to deploy function rnn-serving-1, configs/knative_workloads/rnn_serving.yaml: exit status 1
Creating service 'rnn-serving-1' in namespace 'default':

  0.066s The Route is still working to reflect the latest desired specification.
  0.292s Configuration "rnn-serving-1" is waiting for a Revision to become ready.
 11.161s Revision "rnn-serving-1-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
 11.222s Configuration "rnn-serving-1" does not have any ready Revision.
Error: RevisionFailed: Revision "rnn-serving-1-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
Run 'kn --help' for usage

INFO[0013] Deployed function rnn-serving-1
WARN[0013] Failed to deploy function pyaes-0, configs/knative_workloads/pyaes.yaml: exit status 1
Creating service 'pyaes-0' in namespace 'default':

  0.078s The Route is still working to reflect the latest desired specification.
  0.358s Configuration "pyaes-0" is waiting for a Revision to become ready.
 11.165s Revision "pyaes-0-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
 11.200s Configuration "pyaes-0" does not have any ready Revision.
Error: RevisionFailed: Revision "pyaes-0-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
Run 'kn --help' for usage

INFO[0013] Deployed function pyaes-0
WARN[0013] Failed to deploy function pyaes-1, configs/knative_workloads/pyaes.yaml: exit status 1
Creating service 'pyaes-1' in namespace 'default':

  0.264s The Route is still working to reflect the latest desired specification.
  0.383s Configuration "pyaes-1" is waiting for a Revision to become ready.
 11.300s Revision "pyaes-1-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
 11.419s Configuration "pyaes-1" does not have any ready Revision.
Error: RevisionFailed: Revision "pyaes-1-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
Run 'kn --help' for usage

INFO[0013] Deployed function pyaes-1
WARN[0013] Failed to deploy function rnn-serving-0, configs/knative_workloads/rnn_serving.yaml: exit status 1
Creating service 'rnn-serving-0' in namespace 'default':

  0.269s The Route is still working to reflect the latest desired specification.
  0.399s Configuration "rnn-serving-0" is waiting for a Revision to become ready.
 11.431s Revision "rnn-serving-0-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
 11.480s Configuration "rnn-serving-0" does not have any ready Revision.
Error: RevisionFailed: Revision "rnn-serving-0-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
Run 'kn --help' for usage

INFO[0013] Deployed function rnn-serving-0
WARN[0013] Failed to deploy function helloworld-0, configs/knative_workloads/helloworld.yaml: exit status 1
Creating service 'helloworld-0' in namespace 'default':

  0.380s The Route is still working to reflect the latest desired specification.
  0.460s ...
  0.514s Configuration "helloworld-0" is waiting for a Revision to become ready.
 11.463s Revision "helloworld-0-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
 11.522s Configuration "helloworld-0" does not have any ready Revision.
Error: RevisionFailed: Revision "helloworld-0-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
Run 'kn --help' for usage

INFO[0013] Deployed function helloworld-0
WARN[0026] Failed to deploy function rnn-serving-2, configs/knative_workloads/rnn_serving.yaml: exit status 1
Creating service 'rnn-serving-2' in namespace 'default':

  0.047s The Route is still working to reflect the latest desired specification.
  0.064s Configuration "rnn-serving-2" is waiting for a Revision to become ready.
 11.094s Revision "rnn-serving-2-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
 11.148s Configuration "rnn-serving-2" does not have any ready Revision.
Error: RevisionFailed: Revision "rnn-serving-2-00001" failed with message: Unable to fetch image "crccheck/hello-world:latest": failed to resolve image to digest: Get "https://index.docker.io/v2/": context deadline exceeded.
Run 'kn --help' for usage

INFO[0026] Deployed function rnn-serving-2
INFO[0026] Deployment finished
```

Also, when tried to run the deployer application again, it just shows function deployed, but doesnt seem to be deployed
```
nirav@984fee0096e3:~/vhive$ ./examples/deployer/deployer
INFO[0000] Deployed function rnn-serving-1
INFO[0000] Deployed function pyaes-1
INFO[0000] Deployed function pyaes-0
INFO[0000] Deployed function rnn-serving-0
INFO[0000] Deployed function helloworld-0
INFO[0000] Deployed function rnn-serving-2
INFO[0000] Deployment finished
```
Running invoker just generates empty report
```
nirav@984fee0096e3:~/vhive$ ./examples/invoker/invoker
INFO[2022-02-11T03:50:39.965873525-08:00] Reading the endpoints from the file: endpoints.json
INFO[2022-02-11T03:50:44.966589059-08:00] Issued / completed requests: 5, 0
INFO[2022-02-11T03:50:44.966721266-08:00] Real / target RPS: 0.00 / 1
INFO[2022-02-11T03:50:44.966738443-08:00] Experiment finished!
INFO[2022-02-11T03:50:44.966754156-08:00] The measured latencies are saved in rps0.00_lat.csv
```
rps.0.00_lat.csv is empty here.

**To Reproduce**
./examples/deployer/deployer

**Expected behavior**
functions should get deployed correctly.

**Logs**
nothing relevant in /tmp/vhive-logs/containerd.stderr

**Notes**
Running on 3 Linux server running Ubuntu 20.04

**Comment by harshitgarg22** [comment]
Does `docker pull hello-world` work for you?
Also since you mention being behind a proxy in #475, please check if this can solve your problem: https://stackoverflow.com/questions/23111631/cannot-download-docker-images-behind-a-proxy.

**Comment by niravnshah** [comment]
If i install docker separately, yes, i can do docker pull hello-world after logging into docker after doing "docker login --username". However, as far as i see, the scripts/cloudlab/setup_node.sh doesnt install docker, right? Also, when i tried to install docker separately and then run scripts/cloudlab/setup_node.sh, i could not get the master node kubeadm create cluster correctly. Could be due to this error
`Unable to register node "984fee0096e3.jf.intel.com" with API server: Post "https://10.23.153.129:6443/api/v1/nodes": dial tcp 10.23.153.129:6443: connect: connection refused`
I dont get any such errors when i follow the scripts/cloudlab/setup_node.sh script without explicitly installing docker.

**Comment by niravnshah** [comment]
> Please note that docker and containerd cannot be installed together. Do not install docker on a vHive host.

Thanks for the confirmation @ustiugov. 

@harshitgarg22, i hope your concern about docker is resolved.

**Comment by niravnshah** [comment]
Hi,

the issue with deployer is fixed by following...
`kubectl edit configmap config-deployment -n knative-serving`
add following line under data (dont edit the existing line under _example):
`  registries-skipping-tag-resolving: kind.local,ko.local,dev.local,index.docker.io`

Now, the example services and deployments are created alright, however, i am still not able to successfully invoke them.
Anything to do with the hostnames mentioned in the endpoint.json e.g.
```
nirav@984fee0096e3:~/vhive$ cat endpoints.json
[
        {
                "hostname": "helloworld-0.default.192.168.1.240.sslip.io",
                "eventing": false,
                "matchers": null
        },
        {
                "hostname": "pyaes-0.default.192.168.1.240.sslip.io",
                "eventing": false,
                "matchers": null
        },
        {
                "hostname": "pyaes-1.default.192.168.1.240.sslip.io",
                "eventing": false,
                "matchers": null
        },
        {
                "hostname": "rnn-serving-0.default.192.168.1.240.sslip.io",
                "eventing": false,
                "matchers": null
        },
        {
                "hostname": "rnn-serving-1.default.192.168.1.240.sslip.io",
                "eventing": false,
                "matchers": null
        },
        {
                "hostname": "rnn-serving-2.default.192.168.1.240.sslip.io",
                "eventing": false,
                "matchers": null
        }

```
```
nirav@984fee0096e3:~/vhive$ kubectl get services -o wide
NAME                          TYPE           CLUSTER-IP       EXTERNAL-IP                                            PORT(S)                                      AGE   SELECTOR
helloworld-0                  ExternalName   <none>           knative-local-gateway.istio-system.svc.cluster.local   80/TCP                                       45m   <none>
helloworld-0-00001            ClusterIP      10.98.52.83      <none>                                                 81/TCP                                       45m   <none>
helloworld-0-00001-private    ClusterIP      10.99.105.49     <none>                                                 80/TCP,9090/TCP,9091/TCP,8022/TCP,8013/TCP   45m   serving.knative.dev/revisionUID=a356db3c-8956-4f93-a582-268a94003edb
kubernetes                    ClusterIP      10.96.0.1        <none>                                                 443/TCP                                      75m   <none>
pyaes-0                       ExternalName   <none>           knative-local-gateway.istio-system.svc.cluster.local   80/TCP                                       45m   <none>
pyaes-0-00001                 ClusterIP      10.97.34.194     <none>                                                 81/TCP                                       45m   <none>
pyaes-0-00001-private         ClusterIP      10.99.104.144    <none>                                                 80/TCP,9090/TCP,9091/TCP,8022/TCP,8013/TCP   45m   serving.knative.dev/revisionUID=07f1d98a-299d-4f1d-9aaf-6f4c31ad1b8b
pyaes-1                       ExternalName   <none>           knative-local-gateway.istio-system.svc.cluster.local   80/TCP                                       45m   <none>
pyaes-1-00001                 ClusterIP      10.101.165.178   <none>                                                 81/TCP                                       45m   <none>
pyaes-1-00001-private         ClusterIP      10.107.163.22    <none>                                                 80/TCP,9090/TCP,9091/TCP,8022/TCP,8013/TCP   45m   serving.knative.dev/revisionUID=c9b7162e-728d-4797-99ab-74a0759051c9
rnn-serving-0                 ExternalName   <none>           knative-local-gateway.istio-system.svc.cluster.local   80/TCP                                       45m   <none>
rnn-serving-0-00001           ClusterIP      10.104.203.117   <none>                                                 81/TCP                                       45m   <none>
rnn-serving-0-00001-private   ClusterIP      10.108.27.71     <none>                                                 80/TCP,9090/TCP,9091/TCP,8022/TCP,8013/TCP   45m   serving.knative.dev/revisionUID=3fd80899-630e-41f1-8d77-d9b606cc62ec
rnn-serving-1                 ExternalName   <none>           knative-local-gateway.istio-system.svc.cluster.local   80/TCP                                       44m   <none>
rnn-serving-1-00001           ClusterIP      10.96.87.177     <none>                                                 81/TCP                                       45m   <none>
rnn-serving-1-00001-private   ClusterIP      10.108.61.203    <none>                                                 80/TCP,9090/TCP,9091/TCP,8022/TCP,8013/TCP   45m   serving.knative.dev/revisionUID=8e5090ae-bccc-402d-85e0-86f50237eeef
rnn-serving-2                 ExternalName   <none>           knative-local-gateway.istio-system.svc.cluster.local   80/TCP                                       45m   <none>
rnn-serving-2-00001           ClusterIP      10.100.167.121   <none>                                                 81/TCP                                       45m   <none>
rnn-serving-2-00001-private   ClusterIP      10.99.187.237    <none>                                                 80/TCP,9090/TCP,9091/TCP,8022/TCP,8013/TCP   45m   serving.knative.dev/revisionUID=d2eea7bf-0572-4a09-a425-a5b8ba16d2f5

```
```
nirav@984fee0096e3:~/vhive$ kubectl describe service helloworld-0
Name:              helloworld-0
Namespace:         default
Labels:            serving.knative.dev/route=helloworld-0
                   serving.knative.dev/service=helloworld-0
Annotations:       serving.knative.dev/creator: kubernetes-admin
                   serving.knative.dev/lastModifier: kubernetes-admin
Selector:          <none>
Type:              ExternalName
IP Families:       <none>
IP:
IPs:               <none>
External Name:     knative-local-gateway.istio-system.svc.cluster.local
Port:              http2  80/TCP
TargetPort:        80/TCP
Endpoints:         <none>
Session Affinity:  None
Events:            <none>
nirav@984fee0096e3:~/vhive$ kubectl describe service helloworld-0-00001
Name:              helloworld-0-00001
Namespace:         default
Labels:            app=helloworld-0-00001
                   networking.internal.knative.dev/serverlessservice=helloworld-0-00001
                   networking.internal.knative.dev/serviceType=Public
                   serving.knative.dev/configuration=helloworld-0
                   serving.knative.dev/configurationGeneration=1
                   serving.knative.dev/configurationUID=6cb743bc-5c12-46a9-9d18-9e303a179f6e
                   serving.knative.dev/revision=helloworld-0-00001
                   serving.knative.dev/revisionUID=a356db3c-8956-4f93-a582-268a94003edb
                   serving.knative.dev/service=helloworld-0
                   serving.knative.dev/serviceUID=13e16d96-2937-4094-b476-24f303feb54d
Annotations:       autoscaling.knative.dev/class: kpa.autoscaling.knative.dev
                   autoscaling.knative.dev/target: 1
                   serving.knative.dev/creator: kubernetes-admin
Selector:          <none>
Type:              ClusterIP
IP Families:       <none>
IP:                10.98.52.83
IPs:               10.98.52.83
Port:              http2  81/TCP
TargetPort:        8013/TCP
Endpoints:         192.168.2.5:8013
Session Affinity:  None
Events:            <none>
nirav@984fee0096e3:~/vhive$ kubectl describe service helloworld-0-00001-private
Name:              helloworld-0-00001-private
Namespace:         default
Labels:            app=helloworld-0-00001
                   networking.internal.knative.dev/serverlessservice=helloworld-0-00001
                   networking.internal.knative.dev/serviceType=Private
                   serving.knative.dev/configuration=helloworld-0
                   serving.knative.dev/configurationGeneration=1
                   serving.knative.dev/configurationUID=6cb743bc-5c12-46a9-9d18-9e303a179f6e
                   serving.knative.dev/revision=helloworld-0-00001
                   serving.knative.dev/revisionUID=a356db3c-8956-4f93-a582-268a94003edb
                   serving.knative.dev/service=helloworld-0
                   serving.knative.dev/serviceUID=13e16d96-2937-4094-b476-24f303feb54d
Annotations:       autoscaling.knative.dev/class: kpa.autoscaling.knative.dev
                   autoscaling.knative.dev/target: 1
                   serving.knative.dev/creator: kubernetes-admin
Selector:          serving.knative.dev/revisionUID=a356db3c-8956-4f93-a582-268a94003edb
Type:              ClusterIP
IP Families:       <none>
IP:                10.99.105.49
IPs:               10.99.105.49
Port:              http2  80/TCP
TargetPort:        8013/TCP
Endpoints:         <none>
Port:              http-autometric  9090/TCP
TargetPort:        http-autometric/TCP
Endpoints:         <none>
Port:              http-usermetric  9091/TCP
TargetPort:        http-usermetric/TCP
Endpoints:         <none>
Port:              http-queueadm  8022/TCP
TargetPort:        8022/TCP
Endpoints:         <none>
Port:              http2-istio  8013/TCP
TargetPort:        8013/TCP
Endpoints:         <none>
Session Affinity:  None
Events:            <none>

```
```
nirav@984fee0096e3:~/vhive$ kubectl get deployments -o wide
NAME                             READY   UP-TO-DATE   AVAILABLE   AGE   CONTAINERS                   IMAGES                                                                                                                                                      SELECTOR
helloworld-0-00001-deployment    0/0     0            0           45m   user-container,queue-proxy   crccheck/hello-world:latest,docker.io/hyhe/queue-39be6f1d08a095bd076a71d288d295b6@sha256:e78d4bfc0710ba603bb64aacb2e98aa26d4674a94b67d4633edf4e580725d3aa   serving.knative.dev/revisionUID=a356db3c-8956-4f93-a582-268a94003edb
pyaes-0-00001-deployment         0/0     0            0           45m   user-container,queue-proxy   crccheck/hello-world:latest,docker.io/hyhe/queue-39be6f1d08a095bd076a71d288d295b6@sha256:e78d4bfc0710ba603bb64aacb2e98aa26d4674a94b67d4633edf4e580725d3aa   serving.knative.dev/revisionUID=07f1d98a-299d-4f1d-9aaf-6f4c31ad1b8b
pyaes-1-00001-deployment         0/0     0            0           45m   user-container,queue-proxy   crccheck/hello-world:latest,docker.io/hyhe/queue-39be6f1d08a095bd076a71d288d295b6@sha256:e78d4bfc0710ba603bb64aacb2e98aa26d4674a94b67d4633edf4e580725d3aa   serving.knative.dev/revisionUID=c9b7162e-728d-4797-99ab-74a0759051c9
rnn-serving-0-00001-deployment   0/0     0            0           45m   user-container,queue-proxy   crccheck/hello-world:latest,docker.io/hyhe/queue-39be6f1d08a095bd076a71d288d295b6@sha256:e78d4bfc0710ba603bb64aacb2e98aa26d4674a94b67d4633edf4e580725d3aa   serving.knative.dev/revisionUID=3fd80899-630e-41f1-8d77-d9b606cc62ec
rnn-serving-1-00001-deployment   0/0     0            0           45m   user-container,queue-proxy   crccheck/hello-world:latest,docker.io/hyhe/queue-39be6f1d08a095bd076a71d288d295b6@sha256:e78d4bfc0710ba603bb64aacb2e98aa26d4674a94b67d4633edf4e580725d3aa   serving.knative.dev/revisionUID=8e5090ae-bccc-402d-85e0-86f50237eeef
rnn-serving-2-00001-deployment   0/0     0            0           44m   user-container,queue-proxy   crccheck/hello-world:latest,docker.io/hyhe/queue-39be6f1d08a095bd076a71d288d295b6@sha256:e78d4bfc0710ba603bb64aacb2e98aa26d4674a94b67d4633edf4e580725d3aa   serving.knative.dev/revisionUID=d2eea7bf-0572-4a09-a425-a5b8ba16d2f5

```
```
nirav@984fee0096e3:~/vhive$ kubectl describe deployment helloworld-0-00001-deployment
Name:                   helloworld-0-00001-deployment
Namespace:              default
CreationTimestamp:      Mon, 14 Feb 2022 01:06:45 -0800
Labels:                 app=helloworld-0-00001
                        service.istio.io/canonical-name=helloworld-0
                        service.istio.io/canonical-revision=helloworld-0-00001
                        serving.knative.dev/configuration=helloworld-0
                        serving.knative.dev/configurationGeneration=1
                        serving.knative.dev/configurationUID=6cb743bc-5c12-46a9-9d18-9e303a179f6e
                        serving.knative.dev/revision=helloworld-0-00001
                        serving.knative.dev/revisionUID=a356db3c-8956-4f93-a582-268a94003edb
                        serving.knative.dev/service=helloworld-0
                        serving.knative.dev/serviceUID=13e16d96-2937-4094-b476-24f303feb54d
Annotations:            autoscaling.knative.dev/target: 1
                        deployment.kubernetes.io/revision: 1
                        serving.knative.dev/creator: kubernetes-admin
Selector:               serving.knative.dev/revisionUID=a356db3c-8956-4f93-a582-268a94003edb
Replicas:               0 desired | 0 updated | 0 total | 0 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  0 max unavailable, 25% max surge
Pod Template:
  Labels:       app=helloworld-0-00001
                service.istio.io/canonical-name=helloworld-0
                service.istio.io/canonical-revision=helloworld-0-00001
                serving.knative.dev/configuration=helloworld-0
                serving.knative.dev/configurationGeneration=1
                serving.knative.dev/configurationUID=6cb743bc-5c12-46a9-9d18-9e303a179f6e
                serving.knative.dev/revision=helloworld-0-00001
                serving.knative.dev/revisionUID=a356db3c-8956-4f93-a582-268a94003edb
                serving.knative.dev/service=helloworld-0
                serving.knative.dev/serviceUID=13e16d96-2937-4094-b476-24f303feb54d
  Annotations:  autoscaling.knative.dev/target: 1
                serving.knative.dev/creator: kubernetes-admin
  Containers:
   user-container:
    Image:      crccheck/hello-world:latest
    Port:       50051/TCP
    Host Port:  0/TCP
    Environment:
      GUEST_PORT:       50051
      GUEST_IMAGE:      ghcr.io/ease-lab/helloworld:var_workload
      PORT:             50051
      K_REVISION:       helloworld-0-00001
      K_CONFIGURATION:  helloworld-0
      K_SERVICE:        helloworld-0
    Mounts:             <none>
   queue-proxy:
    Image:       docker.io/hyhe/queue-39be6f1d08a095bd076a71d288d295b6@sha256:e78d4bfc0710ba603bb64aacb2e98aa26d4674a94b67d4633edf4e580725d3aa
    Ports:       8022/TCP, 9090/TCP, 9091/TCP, 8013/TCP
    Host Ports:  0/TCP, 0/TCP, 0/TCP, 0/TCP
    Requests:
      cpu:      25m
    Readiness:  http-get http://:8013/ delay=0s timeout=1s period=1s #success=1 #failure=3
    Environment:
      SERVING_NAMESPACE:                 default
      SERVING_SERVICE:                   helloworld-0
      SERVING_CONFIGURATION:             helloworld-0
      SERVING_REVISION:                  helloworld-0-00001
      QUEUE_SERVING_PORT:                8013
      CONTAINER_CONCURRENCY:             0
      REVISION_TIMEOUT_SECONDS:          300
      SERVING_POD:                        (v1:metadata.name)
      SERVING_POD_IP:                     (v1:status.podIP)
      SERVING_LOGGING_CONFIG:
      SERVING_LOGGING_LEVEL:
      SERVING_REQUEST_LOG_TEMPLATE:      {"httpRequest": {"requestMethod": "{{.Request.Method}}", "requestUrl": "{{js .Request.RequestURI}}", "requestSize": "{{.Request.ContentLength}}", "status": {{.Response.Code}}, "responseSize": "{{.Response.Size}}", "userAgent": "{{js .Request.UserAgent}}", "remoteIp": "{{js .Request.RemoteAddr}}", "serverIp": "{{.Revision.PodIP}}", "referer": "{{js .Request.Referer}}", "latency": "{{.Response.Latency}}s", "protocol": "{{.Request.Proto}}"}, "traceId": "{{index .Request.Header "X-B3-Traceid"}}"}
      SERVING_ENABLE_REQUEST_LOG:        false
      SERVING_REQUEST_METRICS_BACKEND:   prometheus
      TRACING_CONFIG_BACKEND:            none
      TRACING_CONFIG_ZIPKIN_ENDPOINT:
      TRACING_CONFIG_DEBUG:              false
      TRACING_CONFIG_SAMPLE_RATE:        0.1
      USER_PORT:                         50051
      SYSTEM_NAMESPACE:                  knative-serving
      METRICS_DOMAIN:                    knative.dev/internal/serving
      SERVING_READINESS_PROBE:           {"tcpSocket":{"port":50051,"host":"127.0.0.1"},"successThreshold":1}
      ENABLE_PROFILING:                  false
      SERVING_ENABLE_PROBE_REQUEST_LOG:  false
      METRICS_COLLECTOR_ADDRESS:
      CONCURRENCY_STATE_ENDPOINT:
      CONCURRENCY_STATE_TOKEN_PATH:      /var/run/secrets/tokens/state-token
      HOST_IP:                            (v1:status.hostIP)
      ENABLE_HTTP2_AUTO_DETECTION:       false
    Mounts:                              <none>
  Volumes:                               <none>
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  <none>
NewReplicaSet:   helloworld-0-00001-deployment-54dc75b5dc (0/0 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  48m   deployment-controller  Scaled up replica set helloworld-0-00001-deployment-54dc75b5dc to 1
  Normal  ScalingReplicaSet  47m   deployment-controller  Scaled down replica set helloworld-0-00001-deployment-54dc75b5dc to 0

```

**Comment by niravnshah** [comment]
Above mentioned issue of invoker not getting any results is solved by adding ".sslip.io" to no_proxy environment variable

**Comment by niravnshah** [comment]
the issue with deployer is fixed by following...
`kubectl edit configmap config-deployment -n knative-serving`
add following line under data (dont edit the existing line under _example):
`registries-skipping-tag-resolving: kind.local,ko.local,dev.local,index.docker.io`

issue of invoker not getting any results is solved by adding ".sslip.io" to no_proxy environment variable

**Comment by ustiugov** [comment]
great to hear that you find a solution and thank you for reporting back.

**Comment by catchabhisek** [comment]
@niravnshah, I am facing the issue while invoking the sample examples behind a proxy. You said that you have solved it by adding ".sslip.io" to no_proxy environment variable. I tried that out but it didn't work for me. May I know if you have updated the kubelet  or containerd service unit file with this environment variable? 

Currently, this is the deployment status and output of invoker script:

> abhisek@awsum:~$ kubectl get service -o wide
NAME                         TYPE           CLUSTER-IP       EXTERNAL-IP                                            PORT(S)                                      AGE    SELECTOR
helloworld-0                 ExternalName   <none>           knative-local-gateway.istio-system.svc.cluster.local   80/TCP                                       150m   <none>
helloworld-0-00001           ClusterIP      10.105.227.22    <none>                                                 81/TCP                                       150m   <none>
helloworld-0-00001-private   ClusterIP      10.100.234.205   <none>                                                 80/TCP,9090/TCP,9091/TCP,8022/TCP,8013/TCP   150m   serving.knative.dev/revisionUID=96a460e9-6e4f-40f0-b466-d84ff8bedb8d
kubernetes                   ClusterIP      10.96.0.1        <none>                                                 443/TCP                                      30h    <none>

> abhisek@awsum:~$ kubectl get revision -o wide
NAME                 CONFIG NAME    K8S SERVICE NAME   GENERATION   READY   REASON   ACTUAL REPLICAS   DESIRED REPLICAS
helloworld-0-00001   helloworld-0                      1            True             0                 0

> abhisek@awsum:~$ kubectl get configuration -o wide
NAME           LATESTCREATED        LATESTREADY          READY   REASON
helloworld-0   helloworld-0-00001   helloworld-0-00001   True    

> abhisek@awsum:~$ kubectl get route -o wide
NAME           URL                                       READY   REASON
helloworld-0   http://helloworld-0.default.example.com   True    

> abhisek@awsum:~$ kubectl get deployments -o wide
NAME                            READY   UP-TO-DATE   AVAILABLE   AGE    CONTAINERS                   IMAGES                                                                                                                                                      SELECTOR
helloworld-0-00001-deployment   0/0     0            0           151m   user-container,queue-proxy   crccheck/hello-world:latest,docker.io/hyhe/queue-39be6f1d08a095bd076a71d288d295b6@sha256:e78d4bfc0710ba603bb64aacb2e98aa26d4674a94b67d4633edf4e580725d3aa   serving.knative.dev/revisionUID=96a460e9-6e4f-40f0-b466-d84ff8bedb8d

> abhisek@awsum:~$ kubectl get pods -o wide
No resources found in default namespace.

> abhisek@awsum:~$ env | grep no_proxy
no_proxy=127.0.0.1,192.168.122.60,helloworld-0.default.192.168.1.240.sslip.io

> abhisek@awsum:~/vhive$ ./examples/invoker/invoker
INFO[2022-03-21T11:58:56.298149483Z] Reading the endpoints from the file: endpoints.json 
WARN[2022-03-21T11:58:56.309603911Z] Failed to invoke helloworld-0.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unimplemented desc =  
WARN[2022-03-21T11:58:57.305527085Z] Failed to invoke helloworld-0.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unimplemented desc =  
WARN[2022-03-21T11:58:58.305060610Z] Failed to invoke helloworld-0.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unimplemented desc =  
WARN[2022-03-21T11:58:59.306603123Z] Failed to invoke helloworld-0.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unimplemented desc =  
WARN[2022-03-21T11:59:00.304662214Z] Failed to invoke helloworld-0.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unimplemented desc =  
INFO[2022-03-21T11:59:01.298946038Z] Issued / completed requests: 5, 0            
INFO[2022-03-21T11:59:01.299015040Z] Real / target RPS: 0.00 / 1                  
INFO[2022-03-21T11:59:01.299028662Z] Experiment finished!                         
INFO[2022-03-21T11:59:01.299042330Z] The measured latencies are saved in rps0.00_lat.csv

**Comment by aditya2803** [comment]
@catchabhisek Were you able to fix the issue ? I am stuck in a similar position

**Comment by ustiugov** [comment]
@aditya2803 the proxy problem should be solved with this #481. Before it's merged, you can apply this fix locally in your code.

**Comment by aditya2803** [comment]
Hi @ustiugov, I got the deployer problem fixed by the above solution. However, my problem is now with the invoker function. The output of the invoker function looks like below:

```
INFO[2022-05-20T17:14:54.731680096Z] Reading the endpoints from the file: endpoints.json
WARN[2022-05-20T17:14:54.815971522Z] Failed to invoke helloworld-0.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unavailable desc = upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: delayed connect error: 113
WARN[2022-05-20T17:14:55.808471359Z] Failed to invoke pyaes-0.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unavailable desc = upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: delayed connect error: 113
WARN[2022-05-20T17:14:56.805964788Z] Failed to invoke pyaes-1.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unavailable desc = upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: delayed connect error: 113
WARN[2022-05-20T17:14:57.807574347Z] Failed to invoke rnn-serving-0.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unavailable desc = upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: delayed connect error: 113
WARN[2022-05-20T17:14:58.804885784Z] Failed to invoke rnn-serving-1.default.192.168.1.240.sslip.io:80, err=rpc error: code = Unavailable desc = upstream connect error or disconnect/reset before headers. reset reason: connection failure, transport failure reason: delayed connect error: 113
INFO[2022-05-20T17:14:59.732903943Z] Issued / completed requests: 5, 0
INFO[2022-05-20T17:14:59.732988112Z] Real / target RPS: 0.00 / 1
INFO[2022-05-20T17:14:59.733001968Z] Experiment finished!
INFO[2022-05-20T17:14:59.733011886Z] The measured latencies are saved in rps0.00_lat.csv
```
I am running vHive in the stock-only mode for now. And all services, routes, deployments, revisions and pods are running as expected.
Any idea about this ?
